File: frontend/school-portal/src/schemas/finance_schema.ts
export interface Fund {
  id: string;
  name: string;
  short_name?: string;
  receipt_series_prefix?: string;
  receipt_number_start: number;
  current_receipt_number: number;
  school_id: string;
  installment_type?: string;
}

export interface FundCreate {
  name: string;
  short_name?: string;
  receipt_series_prefix?: string;
  receipt_number_start: number;
  school_id: string;
  installment_type?: string;
}

export interface FundUpdate {
  name?: string;
  short_name?: string;
  receipt_series_prefix?: string;
  receipt_number_start?: number;
  installment_type?: string;
}

export interface Fee {
  id: string;
  fee_name: string;
  fee_short_name?: string;
  fund_id: string;
  school_id: string;
}

export interface FeeCreate {
  fee_name: string;
  fee_short_name?: string;
  fund_id: string;
  school_id: string;
}

export interface FeeUpdate {
  fee_name?: string;
  fee_short_name?: string;
  fund_id?: string;
}

export interface ClassFee {
  id: string;
  fee_id: string;
  class_id: string;
  fee_name?: string;
  amount: number;
  installment_type?: string;
  academic_year: string;
}

export interface Payment {
  id: string;
  receipt_number: string;
  student_id: string;
  fund_id: string;
  payment_date: string;
  amount_paid: number;
  payment_mode: string;
  transaction_id?: string;
  remarks?: string;
  school_id: string;
  received_by_user_id: string;
}

export interface PaymentCreate {
  student_id: string;
  fund_id: string;
  payment_date: string;
  amount_paid: number;
  payment_mode: string;
  transaction_id?: string;
  remarks?: string;
  school_id: string;
  received_by_user_id: string;
}

export interface Concession {
  id: string;
  student_id: string;
  fee_id: string;
  discount_amount?: number;
  discount_percentage?: number;
  reason?: string;
  academic_year: string;
  school_id: string;
  approved_by_user_id: string;
}

export interface ConcessionCreate {
  student_id: string;
  fee_id: string;
  discount_amount?: number;
  discount_percentage?: number;
  reason?: string;
  academic_year: string;
  school_id: string;
  approved_by_user_id: string;
}

export interface Salary {
  id: string;
  teacher_id: string;
  month: string; // YYYY-MM
  basic_salary: number;
  allowances: number;
  deductions: number;
  net_salary: number;
  payment_date?: string;
  payment_mode?: string;
  school_id: string;
}

export interface SalaryCreate {
  teacher_id: string;
  month: string; // YYYY-MM
  basic_salary: number;
  allowances: number;
  deductions: number;
  net_salary: number;
  payment_date?: string;
  payment_mode?: string;
  school_id: string;
}

// Extended Finance Schema

export interface ClassFeeCreate {
  fee_id: string;
  class_id: string;
  amount: number;
  installment_type?: string;
  academic_year: string;
}

export interface ClassFeeUpdate {
  amount?: number;
  installment_type?: string;
  academic_year?: string;
}

export interface StudentLookup {
  id: string;
  admission_number: string;
  full_name: string;
  class_name: string;
  outstanding_balance: number;
}

export interface TeacherLookup {
  id: string;
  employee_id: string;
  full_name: string;
  department: string;
}

export interface StudentFeeStructure {
  id: string;
  student_id: string;
  class_fee_id: string;
  academic_year: string;
  total_amount: number;
  discount_amount: number;
  final_amount: number;
  amount_paid: number;
  outstanding_amount: number;
}

export interface StudentFeeAssignment {
  academic_year: string;
  create_installments: boolean;
  installment_type?: string;
}

export interface StudentLedger {
  student_id: string;
  student_name: string;
  admission_number: string;
  class_name: string;
  academic_year: string;
  total_expected: number;
  total_paid: number;
  total_outstanding: number;
  fee_structures: any[];
  payments: any[];
}

export interface StudentOutstanding {
  student_id: string;
  total_outstanding: number;
  by_fee: any[];
  has_overdue: boolean;
  overdue_count: number;
}

export interface PaymentDetailCreate {
  fee_id: string;
  student_fee_structure_id?: string;
  amount: number;
}

export interface PaymentCreateWithAllocations {
  student_id: string;
  fund_id: string;
  payment_date: string;
  amount_paid: number;
  payment_mode: string;
  transaction_id?: string;
  remarks?: string;
  school_id: string;
  payment_details: PaymentDetailCreate[];
}

export interface BulkClassFeeCreate {
  class_fees: ClassFeeCreate[];
}

export interface BulkPaymentRecord {
  student_id: string;
  fund_id: string;
  amount_paid: number;
  payment_mode: string;
  payment_date: string;
  transaction_id?: string;
  remarks?: string;
  auto_allocate: boolean;
}

export interface BulkPaymentCreate {
  school_id: string;
  payments: BulkPaymentRecord[];
}

export interface BulkConcessionCreate {
  fee_id: string;
  discount_amount?: number;
  discount_percentage?: number;
  reason: string;
  academic_year: string;
  school_id: string;
  student_ids?: string[];
  class_ids?: string[];
}

File: backend/app/schemas/finance_schema.py
from pydantic import BaseModel, Field
import uuid
from typing import Optional, List
from datetime import date
from app.models.enums import InstallmentType, FeeApplicability

# Schema for Fund
class FundBase(BaseModel):
    name: str = Field(..., max_length=100)
    short_name: Optional[str] = Field(None, max_length=20)
    receipt_series_prefix: Optional[str] = Field(None, max_length=10)
    receipt_number_start: int = 1
    installment_type: Optional[str] = Field(None, max_length=20)  # monthly, quarterly, half_yearly, yearly

class FundCreate(FundBase):
    school_id: uuid.UUID

class FundUpdate(FundBase):
    pass

class Fund(FundBase):
    id: uuid.UUID
    school_id: uuid.UUID
    current_receipt_number: int
    installment_type: Optional[str] = None

    class Config:
        orm_mode = True

# Schema for Fee
class FeeBase(BaseModel):
    fee_name: str = Field(..., max_length=100)
    fee_short_name: Optional[str] = Field(None, max_length=20)
    fund_id: uuid.UUID

class FeeCreate(FeeBase):
    school_id: uuid.UUID

class FeeUpdate(BaseModel):
    fee_name: Optional[str] = None
    fee_short_name: Optional[str] = None
    fund_id: Optional[uuid.UUID] = None

class Fee(FeeBase):
    id: uuid.UUID
    school_id: uuid.UUID

    class Config:
        orm_mode = True

# Schema for ClassFee
class ClassFeeBase(BaseModel):
    fee_id: uuid.UUID
    class_id: uuid.UUID
    amount: float
    installment_type: InstallmentType = InstallmentType.ANNUALLY
    fee_applicability: FeeApplicability = FeeApplicability.ALL
    academic_year: str

class ClassFeeCreate(ClassFeeBase):
    pass

class ClassFee(ClassFeeBase):
    id: uuid.UUID

    class Config:
        orm_mode = True

# Schema for Payment
class PaymentBase(BaseModel):
    student_id: uuid.UUID
    fund_id: uuid.UUID
    payment_date: date
    amount_paid: float
    payment_mode: str
    transaction_id: Optional[str] = None
    remarks: Optional[str] = None

class PaymentCreate(PaymentBase):
    school_id: uuid.UUID

class PaymentUpdate(BaseModel):
    payment_date: Optional[date] = None
    amount_paid: Optional[float] = None
    payment_mode: Optional[str] = None
    transaction_id: Optional[str] = None
    remarks: Optional[str] = None
    edit_reason: str = Field(..., min_length=3, description="Reason for editing this payment")

class Payment(PaymentBase):
    id: uuid.UUID
    receipt_number: str
    school_id: uuid.UUID
    received_by_user_id: Optional[uuid.UUID] = None

    class Config:
        orm_mode = True

# Schema for Concession
class ConcessionBase(BaseModel):
    student_id: uuid.UUID
    fee_id: uuid.UUID
    discount_amount: Optional[float] = None
    discount_percentage: Optional[float] = None
    reason: Optional[str] = None
    academic_year: str

class ConcessionCreate(ConcessionBase):
    school_id: uuid.UUID

class Concession(ConcessionBase):
    id: uuid.UUID
    school_id: uuid.UUID
    approved_by_user_id: uuid.UUID

    class Config:
        orm_mode = True

# Schema for Salary
class SalaryBase(BaseModel):
    teacher_id: uuid.UUID
    month: str # YYYY-MM
    basic_salary: float
    allowances: float
    deductions: float
    net_salary: float
    payment_date: Optional[date] = None
    payment_mode: Optional[str] = None

class SalaryCreate(SalaryBase):
    school_id: uuid.UUID

class Salary(SalaryBase):
    id: uuid.UUID
    school_id: uuid.UUID

    class Config:
        orm_mode = True

# Schema for StudentFeeStructure
class StudentFeeStructureBase(BaseModel):
    student_id: uuid.UUID
    class_fee_id: uuid.UUID
    academic_year: str
    total_amount: float
    discount_amount: float = 0
    final_amount: float
    outstanding_amount: float

class StudentFeeStructureCreate(StudentFeeStructureBase):
    pass

class StudentFeeStructure(StudentFeeStructureBase):
    id: uuid.UUID
    amount_paid: float

    class Config:
        orm_mode = True

# Schema for FeeInstallment
class FeeInstallmentBase(BaseModel):
    student_fee_structure_id: uuid.UUID
    installment_number: int
    due_date: date
    amount: float

class FeeInstallmentCreate(FeeInstallmentBase):
    pass

class FeeInstallment(FeeInstallmentBase):
    id: uuid.UUID
    paid_amount: float
    status: str

    class Config:
        orm_mode = True

# Schema for PaymentDetail
class PaymentDetailBase(BaseModel):
    fee_id: uuid.UUID
    student_fee_structure_id: Optional[uuid.UUID] = None
    amount: float

class PaymentDetailCreate(PaymentDetailBase):
    pass

class PaymentDetail(PaymentDetailBase):
    id: uuid.UUID
    payment_id: uuid.UUID

    class Config:
        orm_mode = True

# Enhanced Payment Schema with details
class PaymentWithDetails(Payment):
    payment_details: List[PaymentDetail] = []

    class Config:
        orm_mode = True

# Schema for creating payment with allocation
class PaymentCreateWithDetails(BaseModel):
    student_id: uuid.UUID
    fund_id: uuid.UUID
    payment_date: date
    amount_paid: float
    payment_mode: str
    transaction_id: Optional[str] = None
    remarks: Optional[str] = None
    school_id: uuid.UUID
    payment_details: List[PaymentDetailCreate]

# Alias for consistency
PaymentCreateWithAllocations = PaymentCreateWithDetails

# Lookup schemas
class StudentLookup(BaseModel):
    id: str
    admission_number: str
    full_name: str
    class_name: str
    academic_year: Optional[str] = None
    outstanding_balance: float

class TeacherLookup(BaseModel):
    id: str
    employee_id: str
    full_name: str
    department: str

# Student fee assignment
class StudentFeeAssignment(BaseModel):
    academic_year: str
    create_installments: bool = False
    installment_type: Optional[str] = None

# Student ledger
class StudentLedger(BaseModel):
    student_id: str
    student_name: str
    admission_number: str
    class_name: str
    academic_year: str
    total_expected: float
    total_paid: float
    total_outstanding: float
    fee_structures: List[dict]
    payments: List[dict]
    previous_year_arrears: List[dict] = []
    total_arrears: float = 0

# Student outstanding
class StudentOutstanding(BaseModel):
    student_id: str
    total_outstanding: float
    by_fee: List[dict]
    has_overdue: bool
    overdue_count: int

# ClassFee update
class ClassFeeUpdate(BaseModel):
    amount: Optional[float] = None
    installment_type: Optional[InstallmentType] = None
    fee_applicability: Optional[FeeApplicability] = None
    academic_year: Optional[str] = None

# Bulk operations
class BulkClassFeeCreate(BaseModel):
    class_fees: List[ClassFeeCreate]

class BulkPaymentRecord(BaseModel):
    student_id: uuid.UUID
    fund_id: uuid.UUID
    amount_paid: float
    payment_mode: str
    payment_date: date
    transaction_id: Optional[str] = None
    remarks: Optional[str] = None
    auto_allocate: bool = True

class BulkPaymentCreate(BaseModel):
    school_id: uuid.UUID
    payments: List[BulkPaymentRecord]

class BulkConcessionCreate(BaseModel):
    fee_id: uuid.UUID
    discount_amount: Optional[float] = None
    discount_percentage: Optional[float] = None
    reason: str
    academic_year: str
    school_id: uuid.UUID
    student_ids: Optional[List[str]] = None
    class_ids: Optional[List[uuid.UUID]] = None

# Bulk class fee assignment to all students in a class
class BulkClassFeeAssignment(BaseModel):
    academic_year: str
    create_installments: bool = False
    installment_type: Optional[str] = None
